controllers/selfnoderemediation_controller.go

```diff
@@ -311,6 +321,116 @@ func (r *SelfNodeRemediationReconciler) remediateWithResourceDeletion(snr *v1alp
 	return ctrl.Result{}, nil
 }
 
+func (r *SelfNodeRemediationReconciler) remediateOutOfServiceTaint(snr *v1alpha1.SelfNodeRemediation) (ctrl.Result, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.NodeOutOfServiceVolumeDetach) {
+		err := errors.New("this remediation is disabled")
+		r.logger.Error(err, "Set `NodeOutOfServiceVolumeDetach` feature gate")
+		return ctrl.Result{}, err
+	}
+
+	node, err := r.getNodeFromSnr(snr)
+	if err != nil {
+		r.logger.Error(err, "failed to get node", "node name", snr.Name)
+		return ctrl.Result{}, err
+	}
+
+	if r.isFencingCompleted(snr) {
+		r.logger.Info("fencing completed, cleaning up")
+		if node.Spec.Unschedulable {
+			node.Spec.Unschedulable = false
+			if err := r.Client.Update(context.Background(), node); err != nil {
+				if apiErrors.IsConflict(err) {
+					return ctrl.Result{RequeueAfter: time.Second}, nil
+				}
+				r.logger.Error(err, "failed to mark node as schedulable")
+				return ctrl.Result{}, err
+			}
+		}
+
+		// wait until NoSchedulable taint was removed
+		if utils.TaintExists(node.Spec.Taints, NodeUnschedulableTaint) {
+			return ctrl.Result{RequeueAfter: time.Second}, nil
+		}
+
+		if !r.isResourceDeletionCompleted(node) {
+			isExpired, timeLeft := r.isResourceDeletionExpired(snr)
+			if !isExpired {
+				return ctrl.Result{RequeueAfter: timeLeft}, nil
+			}
+			return ctrl.Result{}, errors.New("Not ready to delete out-of-service taint")
+		}
+
+		if err := r.removeOutOfServiceTaint(node); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if err := r.removeNoExecuteTaint(node); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		if controllerutil.ContainsFinalizer(snr, SNRFinalizer) {
+			if err := r.removeFinalizer(snr); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	r.logger.Info("fencing not completed yet, continuing remediation")
+
+	if !r.isNodeRebootCapable(node) {
+		//use err to trigger exponential backoff
+		return ctrl.Result{}, errors.New("Node is not capable to reboot itself")
+	}
+
+	if !controllerutil.ContainsFinalizer(snr, SNRFinalizer) {
+		return r.addFinalizer(snr)
+	}
+
+	if err = r.addNoExecuteTaint(node); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !node.Spec.Unschedulable || !utils.TaintExists(node.Spec.Taints, NodeUnschedulableTaint) {
+		return r.markNodeAsUnschedulable(node)
+	}
+
+	if snr.Status.TimeAssumedRebooted.IsZero() {
+		//todo this also updates the node but we don't need it
+		return r.updateSnrStatus(node, snr)
+	}
+
+	if r.MyNodeName == node.Name {
+		// we have a problem on this node, reboot!
+		return r.rebootIfNeeded(snr)
+	}
+
+	wasRebooted, timeLeft := r.wasNodeRebooted(snr)
+	if !wasRebooted {
+		return ctrl.Result{RequeueAfter: timeLeft}, nil
+	}
+
+	r.logger.Info("TimeAssumedRebooted is old. The unhealthy node assumed to been rebooted", "node name", node.Name)
+
+	if err = r.addOutOfServiceTaint(node); err != nil {
+		return ctrl.Result{}, err
+	}
+
```

Comment:
Hmm I think we can consider moving the Verification of the resource deletion here (before setting the fencing complete).
I think it'll serve 2 purposes:
1. it'll be more consistent with how we decide fencing is complete (same as Resource deletion - after resource are deleted)
2. It'll be much easier to refactor the code: `remediateWithResourceDeletion` and `remediateOutOfServiceTaint` they can share the same code expect for the removal method:
```
func (r *SelfNodeRemediationReconciler) remediateOutOfServiceTaint(snr *v1alpha1.SelfNodeRemediation) (ctrl.Result, error) {
  return remediateWithResourceRemoval(snr, removeByOutOfServiceTaint)
}

func (r *SelfNodeRemediationReconciler) remediateWithResourceDeletion(snr *v1alpha1.SelfNodeRemediation) (ctrl.Result, error) {
   return remediateWithResourceRemoval(snr, removeByResourceDeletion)
}
```

Reply:
I agree the 2 purpose. Now it's moved here. I updated the current flow in the description.